firmware-management/swupdate-handlers/arm-handler-common.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include <mntent.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include "util.h"
+#include <arm-handler-common.h>
+
+char *str_new(const size_t size)
+{
+    char *dst;
+    if (!(dst = malloc(sizeof dst * size)))
+    {
+        ERROR("%s", "Failed to allocate string");
+        return NULL;
+    }
+    return dst;
+}
+
+void str_delete(char *str)
+{
+    free(str);
+}
+
+char *str_copy(const char *src)
+{
+    if (!src)
+        return NULL;
+
+    char *dst = str_new(strlen(src) + 1);
+    if (!dst)
+    {
+        return NULL;
+    }
+
+    if (!strncpy(dst, src, strlen(src)))
+    {
+        ERROR("%s", "Failed to copy str");
+        str_delete(dst);
+        return NULL;
+    }
+
+    if (dst[strlen(src)] != '\0')
+        dst[strlen(src)] = '\0';
+
+    return dst;
+}
+
+int str_endswith(const char *const substr, const char *const fullstr)
+{
+    const size_t endlen = strlen(fullstr) - strlen(substr);
+    if (endlen <= 0)
```

Comment:
`size_t` is unsigned, so this just means `if (endlen == 0)`, but in that case `fullstr` might still end in `substr` (if they are equal)...

I suggest something like:
```C
int str_endswith(const char *const substr, const char *const fullstr)
{
    const size_t substr_len = strlen(substr);
    const size_t fullstr_len = strlen(fullstr);
    if (substr_len > fullstr_len) return 1;
    const ptrdiff_t endlen = fullstr_len - substr_len;
    return strncmp(fullstr + endlen, substr, substr_len);
}
```

Note1: I used `ptrdiff_t` because it's a signed type that should be big enough to hold the difference of two `size_t`s, even though they're not pointers. IIUC, array subscripts are `int`s so if `endlen` won't fit into an int then `fullstr[endlen]` might not work as intended if `endlen` wouldn't fit into an `int`. There's a good chance that I'm misremembering things (or confusing C with C++) though, and I'm probably being too paranoid - an `int` is almost certainly fine for `endlen`, so
```C
    const size_t substr_len = strlen(substr);
    const size_t fullstr_len = strlen(fullstr);
    if (substr_len > fullstr_len) return 1;
    const int endlen = fullstr_len - substr_len;
    return strncmp(&fullstr[endlen], substr, substr_len);
```
is probably fine too.

Note2: If you don't want to define `substr_len` and `fullstr_len` and use `strlen(substr)` and `strlen(fullstr)` throughout the function then using `strncmp` rather than `strcmp` doesn't really make sense:
```
     return strncmp(&fullstr[endlen], substr, strlen(substr));
```
might as well be
```
    return strcmp(&fullstr[endlen], substr);
```
Given that we're already basing the correctness of the function on the fact that `strlen(&fullstr[endlen]) == strlen(substr)`

Reply:
6066a25d52b66b52f665cebd450130027d5f8714

Reply:
```C
     const size_t endlen = fullstr_len - substr_len;
```
Okay, good :+1:. When I wrote "IIUC, array subscripts are ints so if endlen won't fit into an int then fullstr[endlen] might not work as intended", the "IIUC" qualifier was good because I didn't understand correctly.